src/client/components/catalog/AppCatalog.tsx

```diff
@@ -0,0 +1,316 @@
+    this.category = cat;
+    this.applyResultFilters();
+  }
+
+  render() {
+    const pg = this.state.currentPage;
+    const results = this.state.results.slice(0, PAGE_SIZE * pg + PAGE_SIZE);
+    let rowWrapClassName = '';
+    if (this.state.loading && this.state.results.length) {
+      rowWrapClassName = 'o-30';
+    }
+    return (
+      <div className="mt4">
+        <div className='mt3 flex items-baseline justify-between'>
+          <div className='flex' style={{ minWidth: '30rem' }}>
+            <div className='relative'>
+              <i className='fas fa-search black-30 absolute' style={{ top: '0.65rem', left: '0.5rem' }}></i>
+              <SearchInput onSetVal={(val) => this.handleSearchInput(val)} loading={false} />
+            </div>
+
+            { categoryDropdownView(this) }
+
+            <fieldset className='bn pa0 ml3'>
+              <select className='br2 bn bg-light-gray pa2 black-80' onChange={this.handleChangeTag.bind(this)}>
+                <option value='release'>Released</option>
+                <option value='beta'>In beta</option>
+                <option value='dev'>In development</option>
+              </select>
+            </fieldset>
+          </div>
+
+          { runsSorterView(this) }
+        </div>
+
+        { loadingView(this) }
+
+        <div className={rowWrapClassName} style={{ transition: 'opacity linear 0.1s' }}>
+          { results.map(rowView) }
+        </div>
+
+        { loadMoreButtonView(this) }
+
+      </div>
+    );
+  }
+}
+
+// View for the category filter dropdown
+function categoryDropdownView (component: AppCatalog) {
+  const cats = component.state.categories;
+  if (cats.length === 0) {
+    // No categories available; disable and reduce opacity
+    <fieldset className='bn pa0 ml3'>
+      <select disabled className='o-30 br2 bn bg-light-gray pa2 black-80'>
+      </select>
+    </fieldset>
+  }
+  const optionView = (cat: string) => {
+    return <option key={cat} value={cat}>{formatSnakeCase(cat)}</option>
+  }
+  return (
+    <fieldset className='bn pa0 ml3'>
+      <select
+        className='br2 bn bg-light-gray pa2 black-80'
+        onChange={(ev) => component.handleCategoryChange(ev)}
+      >
+        <option value='any'>Any category</option>
+        { cats.map(optionView) }
+      </select>
+    </fieldset>
+  );
+}
+
+// View for the button to sort results by number of app runs
+function runsSorterView (component: AppCatalog) {
+  if (!component.state.results.length) {
+    // Inactive state
+  return (
+    <div className='black-20'>
+      <span className='dib mr2'> Runs </span>
+      <span className='fa fa-circle black-10'></span>
+    </div>
+  );
+  }
+  return (
+    <div className='b black-70 blue pointer dim' onClick={component.handleClickRuns.bind(component)}>
+      <span className='dib mr2'> Runs </span>
+      <span className={component.runsDesc ? "fa fa-caret-down" : "fa fa-caret-up"}></span>
+    </div>
+  );
+}
+
+// Loading spinner
```

Comment:
do you think we should have one spinner component in the generic folder for reuse?

Reply:
:+1: I extracted it out into components/generic/LoadingSpinner.tsx